src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs

```diff
@@ -137,7 +186,28 @@ protected override void InsertItem(int index, T item)
 
             OnCountPropertyChanged();
             OnIndexerPropertyChanged();
-            OnCollectionChanged(NotifyCollectionChangedAction.Add, item, index);
+
+            if (!_skipCollectionChanged)
+            {
+                OnCollectionChanged(NotifyCollectionChangedAction.Add, item, index);
+            }
+        }
+
+        /// <summary>
+        /// Called by base class Collection&lt;T&gt; when a collection of items is added to list;
+        /// raises a CollectionChanged event to any listeners.
+        /// </summary>
+        protected override void InsertItemsRange(int index, IEnumerable<T> collection)
+        {
+            CheckReentrancy();
+
+            _skipCollectionChanged = true;
+            base.InsertItemsRange(index, collection);
```

Comment:
What happens if someone has overridden `InsertItem` or `InsertItemsRange` to do some validation that throws if someone attempts to add something unwanted to the `ObservableCollection<T>`? Should this be wrapped in a try/finally, so that `_skipCollectionChanged` gets set back to `false`? Something like:

```c#
_skipCollectionChanged = true;
try
{
    base.InsertItemsRange(index, collection);
}
finally
{
    _skipCollectionChanged = false;
}
```

?

Reply:
This is good feedback, and we should make sure the `_skipCollectionChanged` is properly set back to false. We should add some test coverage around this as well to verify it is being covered

Reply:
@justinvp I added the recommended change and created a Test Harness to simulate an `Exception` to see how the code will handle it. The `InsertItemsRange`, `RemoveItemsRange` and `ReplaceItemsRange` implementation in `ObservableCollection<T>` will not be executed so I don't know if this is a valid thing we need to handle if someone Overrides the behavior.

Consider this test harness:

```c#
public class TestHarnessObservableCollection<T> : ObservableCollection<T>
{
    public bool SkipCollectionChanged
    {
        get { return (bool)typeof(ObservableCollection<T>).GetField("_skipCollectionChanged", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(this); }
    }

    protected override void InsertItemsRange(int index, IEnumerable<T> collection)
    {
        base.InsertItemsRange(index, collection);
        throw new Exception();
    }

    protected override void RemoveItemsRange(int index, int count)
    {
        throw new Exception();
    }

    protected override void ReplaceItemsRange(int index, int count, IEnumerable<T> collection)
    {
        throw new Exception();
    }
}
```

When I ran the test the code in `ObservableCollection<T>` where we had the try-finally block never got executed since the Exception happened before that code could ever be executed.

My Question for you since you have seen this in the wild, is there a separate implementation of the Test Harness that you are thinking of that I'm missing?

Reply:
Maybe something like:

```c#
[Fact]
public void Test()
{
    var c = new TestHarnessObservableCollection<string>();
    Assert.False(c.SkipCollectionChanged);
    Assert.Throws<Exception>(() => c.AddRange(new string[1]));
    Assert.False(c.SkipCollectionChanged);
}

public class TestHarnessObservableCollection<T> : ObservableCollection<T>
{
    public bool SkipCollectionChanged
    {
        get { return (bool)typeof(ObservableCollection<T>).GetField("_skipCollectionChanged", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(this); }
    }

    protected override void InsertItem(int index, T item)
    {
        throw new Exception();
    }
}
```

Reply:
Or, perhaps better, something like the following, which makes use of the `NonNullObservableCollection<T>` and doesn't involve looking at private implementation details of the collection:

```c#
[Fact]
public void Test()
{
    int collectionChangedCallCount = 0;
    var c = new NonNullObservableCollection<string>();
    c.CollectionChanged += (_, __) => collectionChangedCallCount++;

    Assert.Throws<ArgumentNullException>(() => c.AddRange(new string[1]));
    Assert.Equal(0, collectionChangedCallCount);

    c.Add("hello");
    Assert.Equal(1, collectionChangedCallCount);
}

internal class NonNullObservableCollection<T> : ObservableCollection<T> where T : class
{
    protected override void InsertItem(int index, T item)
    {
        if (item is null) {
            throw new ArgumentNullException(nameof(item));
        }
        base.InsertItem(index, item);
    }

    protected override void SetItem(int index, T item)
    {
        if (item is null) {
            throw new ArgumentNullException(nameof(item));
        }
        base.SetItem(index, item);
    }
}
```

A similar test could be devised for the `RemoveItemsRange` case.

Reply:
After walking through the code again, I understand your point now and it is 100% valid. To clarify our contract:

```c#
public partial class Collection<T>
{
    public void AddRange(IEnumerable<T> collection);

    public void InsertRange(int index, IEnumerable<T> collection);

    public void RemoveRange(int index, int count);

    public void ReplaceRange(int index, int count, IEnumerable<T> collection);

    // These are override by ObservableCollection<T> to raise the event:

    protected virtual void InsertItemsRange(int index, IEnumerable<T> collection);

    protected virtual void RemoveItemsRange(int index, int count);

    protected virtual void ReplaceItemsRange(int index, int count, IEnumerable<T> collection);
}
```

The only methods in this snippet that can be overriden in a subclass are
* InsertItemsRange
* RemoveItemsRange
* ReplaceItemsRange

But, this omits existing code in the `Collection<T>` implementation that is virtual and may be overriden in a subclass.

Looking at a full implementation of `InsertItemsRange` 
```c#
protected virtual void InsertItemsRange(int index, IEnumerable<T> collection)
{
    if (collection == null)
    {
        ThrowHelper.ThrowArgumentNullException(ExceptionArgument.list);
    }

    if (items.IsReadOnly)
    {
        ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_ReadOnlyCollection);
    }

    if ((uint)index > (uint)items.Count)
    {
        ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.index, ExceptionResource.ArgumentOutOfRange_ListInsert);
    }

    if (GetType() == typeof(Collection<T>) && items is List<T> list)
    {
        list.InsertRange(index, collection);
    }
    else
    {
        foreach (T item in collection)
        {
            InsertItem(index++, item);
        }
    }
}
```

Note that it calls either
* `InsertItem`
* `list.InsertRange`

The contract of `InsertItem` is virual:
`protected virtual void InsertItem(int index, T item)`

Which means someone could override this method to throw the exception which would cause the issues as you mentioned.

@justinvp thanks for clarifying this, now that I see the problem this is important to cover and fix. I'll add test coverage for this and make and update today

Reply:
@justinvp I pushed changes and included a new test class to cover this. Let me know if there is anything else you need to cover this use case.